docker-compose.yml

```diff
@@ -1,306 +1,441 @@
+    restart: unless-stopped
+    environment:
+      - PARALLEL_PREPROCESSORS=ON
+      - STORE_IMAGE_DATA=ON
+    env_file:
+      - ./config/express-common.env
+    group_add:
+      # This should be set in an environment variable or .env
+      - ${DOCKER_GID}
+    volumes:
+      - /var/run/docker.sock:/var/run/docker.sock:ro
+      - user-logs:/var/log/IMAGE
+
+  espnet-tts:
+    profiles: [pegasus, unicorn, default]
+    image: ghcr.io/shared-reality-lab/image-service-espnet-tts:${REGISTRY_TAG}
+    restart: unless-stopped
+    environment:
+      - TORCH_DEVICE=cuda
+    ports:
+      - "8001:80"
+    deploy:
+      resources:
+        reservations:
+          devices:
+          - driver: nvidia
+            capabilities: ["gpu", "utility", "compute"]
+
+  espnet-tts-fr:
+    profiles: [pegasus, unicorn, default]
+    image: ghcr.io/shared-reality-lab/image-service-espnet-tts-fr:${REGISTRY_TAG}
+    restart: unless-stopped
+    environment:
+      - TORCH_DEVICE=cuda
+    deploy:
+      resources:
+        reservations:
+          devices:
+            - driver: nvidia
+              capabilities: ["gpu", "compute", "utility"]
+
+  multilang-support:
+    profiles: [pegasus, unicorn, default]
+    image: ghcr.io/shared-reality-lab/image-service-translation:${REGISTRY_TAG}
+    restart: unless-stopped
+    deploy:
+      resources:
+        reservations:
+          devices:
+            - driver: nvidia
+              capabilities: ["gpu", "compute", "utility"]
+
+  autour-preprocessor:
+    profiles: [pegasus, unicorn, default]
+    image: ghcr.io/shared-reality-lab/image-preprocessor-autour:${REGISTRY_TAG}
+    restart: unless-stopped
+    env_file:
+      - ./config/maps.env
+    labels:
+      ca.mcgill.a11y.image.preprocessor: 1
+      ca.mcgill.a11y.image.port: 5000
+
+  content-categoriser:
+      profiles: [pegasus, unicorn, default]
+      image: ghcr.io/shared-reality-lab/image-preprocessor-content-categoriser:${REGISTRY_TAG}
+      restart: unless-stopped
+      labels:
+        ca.mcgill.a11y.image.preprocessor: 1
+        ca.mcgill.a11y.image.port: 5000
+      deploy:
+        resources:
+          limits:
+            memory: 2g
+
+  graphic-tagger:
+    profiles: [pegasus, unicorn, default]
+    image: ghcr.io/shared-reality-lab/image-preprocessor-graphic-tagger:${REGISTRY_TAG}
+    restart: unless-stopped
+    env_file:
+      - ./config/azure-api.env
+    labels:
+      ca.mcgill.a11y.image.preprocessor: 2
+      ca.mcgill.a11y.image.port: 5000
+
+  object-detection:
+    profiles: [pegasus, unicorn, default]
+    image: ghcr.io/shared-reality-lab/image-preprocessor-object-detection-v8:${REGISTRY_TAG}
+    restart: unless-stopped
+    labels:
+      ca.mcgill.a11y.image.preprocessor: 3
+      ca.mcgill.a11y.image.port: 5000
+    deploy:
+          resources:
+            reservations:
+              devices:
+              - driver: nvidia
+                device_ids: ['1']
```

Comment:
Won't this fail on pegasus, since it only has a single GPU? Need a way to override this based on generic docker-compose for differences in things like GPU hardware (unicorn has two GPUs). Suggest default being the pegasus single card scenario, and an override activated for unicorn, where we need to split specific preprocessors onto second GPU?

Reply:
If we knew that docker-compose would assign them in a reasonable way just always using "all", we wouldn't have to deal with this at all. But I don't know how it does the assignments in that case. So, I assume we want the "default" to be "all" (which should work fine on Pegasus since it only has one GPU) and would be a reasonable default for most external configs. Your approach here makes sense, then, even if it seems messy. The only other way that comes to mind would be to make a docker-compose.override.yml file that only changes the GPU assignments. But then, our standard operating procedure for having people create their own override when testing on unicorn becomes more difficult. Multiple overrides are supported, but would need to be manually specified, or we'd have to create nice helper scripts that do the right thing. Not sure if that is worth it. @JRegimbal suggested another .env variable, but this seems weird to me since it is tightly linked to specific preprocessors. Would that really be cleaner? I'm open to debate on this, or to another alternative... I'm fine with the current solution you've implemented for now, and we can revisit later. In speaking with Juliette, one consideration is what will scale better if we move to something like docker swarm, but I don't know since I've never used swarm...

Reply:
I don't love this, but I'm fine with the two versions in the meantime if there's a comment clearly explaining what's happening.